### PR TITLE
745 move authenticate user method

### PR DESCRIPTION
**Pull request by Gabe-Torres**
# 🔗 Issue
#745 

# ✍️ Description
Changes made for issue 745 included:

1. Moving the `before_action :authenticate_user!`  from the `Organizations::AdopterFosterer::AdopterApplicationsController` and `Organizations::AdopterFosterer::ProfilesController`
into the `ApplicationController`
2. Adding `skip_before_action :authenticate_user!` before the controllers mentioned in the issue 
`static pages`
`errors`
`contacts`
`organizations - home`
`organizations - FAQ`
`organizations - adoptable pets`

and 6 additional controllers
`Organizations::Staff::FosterersController `
`Organizations::Staff::StaffController`
`RootController`
`StatesController`
`Organizations::Staff::ManageFostersController`
`Organizations::Staff::AdoptionApplicationReviewsController`

# 📷 Screenshots/Demos
https://jam.dev/c/ee93ea30-9c83-4a15-a897-8ab05082f935

### CODE REVIEW

**Comment by Gabe-Torres** [comment]
Looks like this pr could use a rebase! Let me know if you'd like me to take care of that or if there are changes you would like me to make 👍🏽

File: app/controllers/organizations/staff/adoption_application_reviews_controller.rb
Comment (kasugaijin) [inline app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: Why would we skip this on the staff scoped actions?

Comment (Gabe-Torres) [reply app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: Initially, my thought process was to comment out the `before_action :authenticate_user!` within the `Organizations::AdopterFosterer::AdopterApplicationsController` and `Organizations::AdopterFosterer::ProfilesController` and see tests fail. But all tests passed. So I moved the `before_action :authenticate_user!` to the `ApplicationController` and I got failing tests. Which made sense to me. I followed the errors and skipped authentication within those controllers. Eventually, all tests passed. Without skipping on the staff scoped actions, I get the following error
`Failure:
Organizations::Staff::StaffControllerTest#test_: authorization #activate should be authorized.  [test/controllers/organizations/staff/staff_controller_test.rb:25]:
Expected #<StaffAccount id: 2649, created_at: "2024-06-01 18:00:05.806002000 +0000", updated_at: "2024-06-01 18:00:05.806002000 +0000", organization_id: 5169, user_id: 4740, deactivated_at: nil> to be authorized with Organizations::StaffAccountPolicy#activate?, and context {}, but no such authorization has been made.
Registered authorizations: none`

Comment (kasugaijin) [reply app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: Hmm interesting. We can take a look in tomorrow meeting. My initial question is why would an authorizations related test be failing for an authentication action.

Comment (Gabe-Torres) [reply app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: That’s a good question I’ll also look into it more to see if I can do something differently

Comment (kasugaijin) [reply app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: Comment here if you find anything! We will discuss I’m on the weekly meeting tomorrow if you are interested in joining that. Ping me in slack for the link. It’s at 8 pm eastern time US.

Comment (Gabe-Torres) [reply app/controllers/organizations/staff/adoption_application_reviews_controller.rb:2]: Will do! I sent a dm for the link

**Comment by Gabe-Torres** [comment]
Hey @kasugaijin !
I believe I made the correct adjustments in my most recent push. Authentication is still skipped in the root controller. I remember it was mentioned that makes sense, but let me know and I can make the necessary changes.

One thing to note:
Authentication is required or assumed when reaching the majority of routes. However, if a user accesses certain routes in the URL like `/staff/fosters` they are greeted with something like `Organization doesn't exist`. Previously, there was no attempt to reroute the user to a login page and they were greeted with a flash error while remaining on the homepage. But now with the added authentication in the application controller, there is an attempt to do so, but it is not quite doing it. The user is still greeted with the same flash error `Organization doesn't exist` and remains on the homepage. I figured the redirect to the login page might be another issue altogether. Which I would be happy to try and tackle! 
<h1>Before without authentication</h1>
 
Screenshot 2024-06-06 at 6 54 06 PM

<h1>After with authentication</h1>

Screenshot 2024-06-06 at 6 57 45 PM

**Comment by Gabe-Torres** [comment]
To summarize the changes: 
sign_in was required for the following controller tests:
`fosterers_controller_test`
`fosters_controller_test`
`staff_controller_test`
`states_controller_test`

and integration test:

`adoption_application_review_test`

**Comment by kasugaijin** [comment]
@Gabe-Torres sorry I missed the notification of your comments above - just seeing them now.

[Video thoughts](https://www.loom.com/share/591959d2ce2446348f983f461027ee66?sid=750fb77f-9ed3-4bf4-b894-f583f9828ea7)

**Comment by Gabe-Torres** [comment]
@kasugaijin No problem! Thank you for the video too. That provides a lot of clarity. 
On my branch, I tried adding the slugs `baja` and `alta` to the endpoint, and the same thing occurred. 
I did a checkout to main to give it a try and a redirect to the organization homepage occurs.

here are some screenshots

<h1>My branch</h1>

Screenshot 2024-06-11 at 5 50 53 PM

<h1>Main branch</h1>

Screenshot 2024-06-11 at 5 54 11 PM

It looks like the `skip_before_action :authenticate_user!` in the `root controller` might have caused the redirect to not occur for some reason. I'll look into this more and take the skip before action out of the root controller.

**Comment by kasugaijin** [comment]
Hmm I think we will need `skip_before_action :authenticate_user!` in the root controller though? Otherwise people will have to be authenticated to access the org root.

**Comment by Gabe-Torres** [comment]
That was my thought too! Not sure why the reroute to the organization page is being cancelled by skipping though

**Comment by kasugaijin** [comment]
Ahh Ok I see what's going on here - it's a bit of ping pong but hopefully my video explains it - let me know if not!

https://www.loom.com/share/204985feda8848e6bb812d294d69bf92?sid=231e7cbc-2730-495b-b641-ae31dfdc6a08

**Comment by Gabe-Torres** [comment]
Thank you for looking into the issue! That does make sense. And I see what's going on. Let me see what I can do 👍🏽

**Comment by Gabe-Torres** [comment]
https://jam.dev/c/3c19d1cc-b5bc-42b2-a296-697885710d60

**Comment by Gabe-Torres** [comment]
https://jam.dev/c/d4e1f6ad-ccc0-4867-b271-5aca10d9bbfc

Here's a video that overviews my recent updates.

It involves creating a module that inherits from Devises Failure App and overwrites the method `scoped_url`. It's the same aside from commenting out one line. `opts[:script_name] = nil`
That particular line was overwriting our `OrganizationMiddleware` class and not allowing it to create the route we needed when a user is not authenticated. With these updates Devise can do its thing and work with the `OrganizationMiddleware` class to create the slug for authenticated routes.

File: lib/authentication_failure_app.rb
Comment (kasugaijin) [inline lib/authentication_failure_app.rb:10]: Ah very nice! Great find. I have not worked with this module before so I learned something new. 

At the risk of writing a novel, could we add just a bit more context in these comments to provide a crystal clear explanation of why this is important? I think what you have is good, but you could just add something like 'I.e., if an unauthenticated user tried to access an authenticated route within an organization's scope, they would be redirected to the sign in page, but the organization scope would be lost in the URL. This fixes that to make sure the Organization's slug is present in the URL' or something like that?

Hopefully this was an interesting journey for you, too! I learned something.

Comment (Gabe-Torres) [reply lib/authentication_failure_app.rb:10]: Will do! Yeah, I enjoyed this issue and learned some new concepts as well. I've used Devise a few times on some personal projects but never with one this large! I'm sure there are other ways to achieve the same end result, but I quite like this method.

I noticed there are some merge conflicts; should I take care of those too?

Comment (Gabe-Torres) [reply lib/authentication_failure_app.rb:18]: I added some more detail on what this module/method is doing. Let me know if it sounds clear enough or if it needs to be trimmed down!

Comment (kasugaijin) [reply lib/authentication_failure_app.rb:18]: Yes that looks good thank you!

**Comment by kasugaijin** [comment]
Ah, something got merged...conflicts :)

**Review by kasugaijin — APPROVED** [review]
LGTM! I am curious how this will behave on the app route when we eventually implement sysadmin role for devs, who will log in outside of the org context. But, my personal preference there is to create brand new devise resource anyway, that is not scoped to any org, so it should not be an issue.

**Comment by Gabe-Torres** [comment]
Merge conflicts! The best part lol🥇 Shouldn't be too hard to resolve! Should I take a go at them?

**Comment by kasugaijin** [comment]
Yes please! Yeh they should be pretty straightforward but let me know if you have any questions

**Comment by Gabe-Torres** [comment]
I sorted out the merge conflicts! Just waiting on another approval now 👍

File: db/schema.rb
Comment (kasugaijin) [inline db/schema.rb:1]: I don’t think you are meant to be adding any schema changes?

**Review by kasugaijin — APPROVED** [review]
Lgtm! It was not so easy after all despite the issue label but good job seeing it through!

**Comment by kasugaijin** [comment]
Ahh one more thing to pass the linter please run rails standard:fix and push up to this pr

**Comment by Gabe-Torres** [comment]
Thank you! From my experience so far, code never seems to be as simple as one would like, but those situations make for the best growth 🎉 
Updated and pushed up the changes after running `rails standard fix`!

**Pull request merged**
